### DISCUSSION TITLE
ESP32 BLE HID device stopped working with MicroPython version 1.20 and up

### DISCUSSION THREAD

**Discussion by Heerkog** [micropython/micropython#14258] [General]
I have written a library with several HID services. The library, however, stopped functioning with MicroPython version 1.20 and up, and I have no idea why. The client just seems to disconnect shortly after it has connected. Neither the MicroPython documentation, nor the BLE examples seem to indicate significant changes that would cause such behavior. Any help would be appreciated in tracking down the issue.

The library can be found at:
https://github.com/Heerkog/MicroPythonBLEHID/blob/master/hid_services.py

A discussion of the issue can be found at:
https://github.com/Heerkog/MicroPythonBLEHID/issues/16

**Comment by jonnor** [comment]
This is for ESP32, is it not? And it is working correctly on MicroPython on 1.19 ? If so, I would recommend trying to bisect the changes between 1.19 and 1.20 to find where it started to break.

**Reply by Heerkog** [reply]
Thank you for your reply.

Correct, it is for ESP32.

I have gone over the changes, but have not been able to pinpoint the cause. I'm suspecting some undocumented changes. Hence me asking for help here.

**Reply by easytarget** [reply]
'Bisecting the changes' means picking a daily build halfway between 1.19 and 1.20. Trying that and if the issue is present there picking a build halfway between 1.19 and it (or if not present, trying one halfway to 1.20), Then keep repeating that, homing in on the error, until you get to the level of individual commits. Usually you can track it down to a single commit.

Which is easy for me to say.. but can be a total slow PITA in reality.

It requires that you can build the ESP32 port since afik. daily builds are not archived in this detail, and a certain agility with Git.

**Comment by shariltumin** [comment]
If I remember correctly, MicroPython makes a transition from ESP-IDF4.4 to ESP-IDF5.0 around MP1.19 to MP1.20.

```
$ strings firmware.bin | grep IDF
[0;32mI (%lu) %s: ESP-IDF:          %s
IDF version : v5.2.2
MicroPython-1.24.0-preview-xtensa-IDFv5.2.2-with-newlib4.3.0
```
 
You can check which ESP-IDF version works for your HID module. ESP-IDF5.x may not be compatible with your code.
